src/plugins/ui_feature/find_and_replace.ts

```diff
@@ -213,47 +213,68 @@ export class FindAndReplacePlugin extends UIPlugin {
+  private replaceMatch(replaceWith: string, selectedMatch: SearchMatch): boolean {
+    if (!this.currentSearchRegex) {
+      return false;
+    }
+
+    const sheetId = this.getters.getActiveSheetId();
+    const cell = this.getters.getCell({ sheetId, ...selectedMatch });
+
+    if (cell?.isFormula && !this.searchOptions.searchFormulas) {
+      return false;
+    }
+
+    const replaceRegex = new RegExp(
+      this.currentSearchRegex.source,
+      this.currentSearchRegex.flags + "g"
+    );
+    const toReplace: string | null = this.getSearchableString({
+      sheetId,
+      col: selectedMatch.col,
+      row: selectedMatch.row,
+    });
+    const newContent = toReplace.replace(replaceRegex, replaceWith);
+    this.dispatch("UPDATE_CELL", {
+      sheetId: this.getters.getActiveSheetId(),
+      col: selectedMatch.col,
+      row: selectedMatch.row,
+      content: newContent,
+    });
+    return true;
+  }
+
   /**
    * Replace the value of the currently selected match
    */
   private replace(replaceWith: string) {
-    if (this.selectedMatchIndex === null || !this.currentSearchRegex) {
+    if (this.selectedMatchIndex === null) {
       return;
     }
-    const matches = this.searchMatches;
-    const selectedMatch = matches[this.selectedMatchIndex];
-    const sheetId = this.getters.getActiveSheetId();
-    const cell = this.getters.getCell({ sheetId, ...selectedMatch });
-    if (cell?.isFormula && !this.searchOptions.searchFormulas) {
-      this.selectNextCell(Direction.next);
-    } else {
-      const replaceRegex = new RegExp(
-        this.currentSearchRegex.source,
-        this.currentSearchRegex.flags + "g"
-      );
-      const toReplace: string | null = this.getSearchableString({
-        sheetId,
-        col: selectedMatch.col,
-        row: selectedMatch.row,
-      });
-      const newContent = toReplace.replace(replaceRegex, replaceWith);
-      this.dispatch("UPDATE_CELL", {
-        sheetId: this.getters.getActiveSheetId(),
-        col: selectedMatch.col,
-        row: selectedMatch.row,
-        content: newContent,
-      });
+
+    const selectedMatch = this.searchMatches[this.selectedMatchIndex];
+    const replaced = this.replaceMatch(replaceWith, selectedMatch);
+
+    if (replaced) {
       this.searchMatches.splice(this.selectedMatchIndex, 1);
       this.selectNextCell(Direction.current);
+    } else {
+      this.selectNextCell(Direction.next);
     }
   }
   /**
    * Apply the replace function to all the matches one time.
    */
   private replaceAll(replaceWith: string) {
-    const matchCount = this.searchMatches.length;
-    for (let i = 0; i < matchCount; i++) {
-      this.replace(replaceWith);
+    for (let i = 0; i < this.searchMatches.length; i++) {
+      const replaced = this.replaceMatch(replaceWith, this.searchMatches[i]);
+      if (replaced) {
+        this.searchMatches.splice(i, 1);
+        i--; // Adjust the index due to splice operation
```

Comment:
surely there is another way to meet the functional requirement than hacking the index inside a loop ...

Reply:
Can you please check this?